src/chunk.c

```diff
@@ -62,31 +58,46 @@ ChunkResult Uncompressed_AddSample(Chunk_t *chunk, Sample *sample) {
         regChunk->base_timestamp = sample->timestamp;
     }
 
-    ChunkGetSampleArray(regChunk)[regChunk->num_samples] = *sample;
+    regChunk->samples[regChunk->num_samples] = *sample;
     regChunk->num_samples++;
 
     return CR_OK;
 }
 
-ChunkIter_t *Uncompressed_NewChunkIterator(Chunk_t *chunk) {
+ChunkIter_t *Uncompressed_NewChunkIterator(Chunk_t *chunk, int options) {
     ChunkIterator *iter = (ChunkIterator *)calloc(1, sizeof(ChunkIterator));
     iter->chunk = chunk;
-    iter->currentIndex = 0;
-    return iter;
+    iter->currentIndex = (options & REVERSE) == 0 ? 0 : iter->chunk->num_samples - 1;
+    iter->options = options;
+    return (ChunkIter_t *)iter;
 }
 
 ChunkResult Uncompressed_ChunkIteratorGetNext(ChunkIter_t *iterator, Sample *sample) {
-    ChunkIterator *iter = iterator;
+    ChunkIterator *iter = (ChunkIterator *)iterator;
     if (iter->currentIndex < iter->chunk->num_samples) {
+        *sample = *ChunkGetSample(iter->chunk, iter->currentIndex);
         iter->currentIndex++;
-        Sample *internalSample = ChunkGetSample(iter->chunk, iter->currentIndex - 1);
-        memcpy(sample, internalSample, sizeof(Sample));
         return CR_OK;
     } else {
         return CR_END;
     }
 }
 
+ChunkResult Uncompressed_ChunkIteratorGetPrev(ChunkIter_t *iterator, Sample *sample) {
+    ChunkIterator *iter = (ChunkIterator *)iterator;
+    if (iter->currentIndex >= 0) {
+        *sample = *ChunkGetSample(iter->chunk, iter->currentIndex);
+        iter->currentIndex--;
+        return CR_OK;
+    } else {
+        return CR_END;
+    }
+}
+
+void FreeChunkIterator(ChunkIter_t *iter, bool freeChunk) {
+  free(iter);
```

Comment:
@ashtul Don't you need to free the chunk if `freeChunk` is enabled?

Reply:
Looking at the rest of the implementation I see that `FREE_TEMP_CHUNK` should be enabled in `iter->options`, so we need to check the flag and free the chunk if enabled.

Reply:
Changed the method since we use a common generic pointer for both compressed and uncompressed iter.
Removed option and placed rev instead.